ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java

```diff
@@ -101,4 +205,468 @@ public Integer getValue() {
+     * This method should be called when a response of a request has been received and processed. This method should
+     * not be called when a request is aborted.
+     * @param latency Request latency.
+     * @param operationType {@link com.github.ambry.router.NonBlockingRouter.RouterOperationType}.
+     */
+    void updateRequestLatency(long latency, NonBlockingRouter.RouterOperationType operationType) {
+      switch (operationType) {
+        case PutBlob:
+          putRequestLatencyMs.update(latency);
+          break;
+        case GetBlobInfo:
+          getBlobInfoRequestLatencyMs.update(latency);
+          break;
+        case GetBlob:
+          getRequestLatencyMs.update(latency);
+          break;
+        case DeleteBlob:
+          deleteRequestLatencyMs.update(latency);
+          break;
+        default:
+          // We simply log the error here and do not add a metrics to record this, as the default case should not happen.
+          logger.warn("Error for unknown RouterOperationType for updating request latency: " + operationType);
+          break;
+      }
+    }
+
+    /**
+     * Update the counter for timedout requests for a given {@link com.github.ambry.router.NonBlockingRouter.RouterOperationType}.
+     * @param operationType {@link com.github.ambry.router.NonBlockingRouter.RouterOperationType}
+     */
+    void updateRequestTimeoutCount(NonBlockingRouter.RouterOperationType operationType) {
+      switch (operationType) {
+        case PutBlob:
+          getBlobInfoRequestTimeoutCount.inc();
+          break;
+        case GetBlobInfo:
+          getBlobInfoRequestTimeoutCount.inc();
+          break;
+        case GetBlob:
+          getRequestTimeoutCount.inc();
+          break;
+        case DeleteBlob:
+          deleteRequestTimeoutCount.inc();
+          break;
+        default:
+          // We simply log the error here and do not add a metrics to record this, as the default case should not happen.
+          logger.warn("Error for unknown RouterOperationType for updating timed-out request count: " + operationType);
+          break;
+      }
+    }
+
+    /**
+     * Count the {@link RouterException} caused by a request (or its response) for a given
+     * {@link NonBlockingRouter.RouterOperationType}. This {@link RouterException} may not be the final
+     * {@link RouterException} accompanied with the {@code Operation}, but such metrics can assist a
+     * root-cause analysis at the {@link DataNodeId} level.
+     * <p/>
+     * This method should be called when a request is either completed or timed out.
+     * @param error The {@link RouterErrorCode} of the exception to be counted.
+     * @param operationType The type of the operation that is associated with the given exception.
+     */
+    void countError(RouterErrorCode error, NonBlockingRouter.RouterOperationType operationType) {
+      if (error == null) {
```

Comment:
Also, why are we handling `error == null` case? I think you are missing the point here. The metrics should be updated only when there is an error. Would rather not see all responsibilities be dumped into the metrics related methods.